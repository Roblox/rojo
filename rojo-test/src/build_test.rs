use std::{
    fs,
    path::Path,
    process::Command,
};

use insta::assert_snapshot_matches;
use tempfile::tempdir;

macro_rules! gen_build_tests {
    ( $($test_name: ident,)* ) => {
        $(
            paste::item! {
                #[test]
                fn [<build_ $test_name>]() {
                    run_build_test(stringify!($test_name));
                }
            }
        )*
    };
}

gen_build_tests! {
    client_in_folder,
    csv_bug_145,
    csv_bug_147,
    csv_in_folder,
    gitkeep,
    module_in_folder,
    plain_gitkeep,
    rbxm_in_folder,
    rbxmx_in_folder,
    server_in_folder,
    txt,
    txt_in_folder,
}

#[test]
fn build_plain_txt() {
    run_build_test("plain.txt");
}

#[test]
fn build_rbxmx_ref() {
    run_build_test("rbxmx_ref.rbxmx");
}

fn run_build_test(test_name: &str) {
    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    let build_test_path = manifest_dir.join("build-tests");
    let working_dir = manifest_dir.parent().unwrap();

    let output_dir = tempdir().expect("couldn't create temporary directory");

    let input_path = build_test_path.join(test_name);
    let output_path = output_dir.path().join(format!("{}.rbxmx", test_name));

    let status = Command::new("cargo")
        .args(&[
            "run", "--quiet", "--",
            "build", input_path.to_str().unwrap(), "-o", output_path.to_str().unwrap(),
        ])
        .current_dir(working_dir)
        .status()
        .expect("Couldn't start Rojo");

    assert!(status.success(), "Rojo did not exit successfully");

    let contents = fs::read_to_string(&output_path)
        .expect("Couldn't read output file");

    assert_snapshot_matches!(test_name, contents);
}