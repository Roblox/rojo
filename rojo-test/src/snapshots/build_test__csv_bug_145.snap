---
created: "2019-08-14T22:42:52.297447700Z"
creator: insta@0.10.0
source: rojo-test/src/build_test.rs
expression: contents
---
<roblox version="4">
  <Item class="Folder" referent="0">
    <Properties>
      <string name="Name">csv_bug_145</string>
    </Properties>
    <Item class="LocalizationTable" referent="1">
      <Properties>
        <string name="Name">normal</string>
        <string name="Contents">[{"key":"Count","example":"A number demonstrating issue 145","source":"3","values":{"es":"7"}}]</string>
      </Properties>
    </Item>
  </Item>
</roblox>